pkg/apis/cloudcredential/v1/credentialsrequest_types.go

```diff
@@ -44,17 +44,55 @@ const (
 	CloudCredOperatorConfigMap = "cloud-credential-operator-config"
 )
 
-// NOTE: Run "make" to regenerate code after modifying this file
+// NOTE: Run "make update" to regenerate code after modifying this file
 
 // CredentialsRequestSpec defines the desired state of CredentialsRequest
 type CredentialsRequestSpec struct {
 	// SecretRef points to the secret where the credentials should be stored once generated.
-	SecretRef corev1.ObjectReference `json:"secretRef"`
+	// +optional
+	SecretRef corev1.ObjectReference `json:"secretRef,omitempty"`
+
+	// StorageDestination specifies the type and location of a resource to which the credentials
+	// information can be stored or linked
+	StorageDestination StorageDestination `json:"storage,omitempty"`
 
 	// ProviderSpec contains the cloud provider specific credentials specification.
 	ProviderSpec *runtime.RawExtension `json:"providerSpec,omitempty"`
 }
 
+const (
+	SecretStorageDestinationType         = "Secret"
+	ServiceAccountStorageDestinationType = "ServiceAccount"
+)
+
+// StorageDestination defines the type and location of a resource to which the credentials
+// information can be stored or linked
+// +union
+type StorageDestination struct {
+	// StorageType contains the type of storage that should be used for the credentials
+	// +unionDiscriminator
+	Type string `json:"type"`
+
+	// SecretStorageDestination specifies a secret in which to store the credentials
+	// +optional
+	Secret *SecretStorageDestination `json:"secret,omitempty"`
+
+	// SecretStorageDestination specifies a serviceaccount to which the credentials will be linked
+	// This is currently only supported on AWS
+	// +optional
+	ServiceAccount *ServiceAccountStorageDestination `json:"serviceAccount,omitempty"`
+}
+
+type SecretStorageDestination struct {
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+}
```

Comment:
IMO we should just stick with ObjectReference here, even if we don't use or honor all the fields.

Reply:
@deads2k how strong is your opinion that `ObjectReference` should not be used?

Reply:
If this is best practice I'd be curious to know.

However we should stay consistent with secretRef which is already ObjectReference. 

Perhaps we could switch it to a type though, as we only would be respecting the Name and Namespace, and API callers would not be impacted?

In any case, would like to see them the same.

Reply:
Just heard in another review that ObjectReference use is discouraged. So can we safely flip secretRef? Should we make them different?

Reply:
I'm not saying we should change `secretRef` but if we are going to create a new `storage` union as part of this work, I'm saying we shouldn't make the _new_ field `ObjectReference`

Reply:
> @deads2k how strong is your opinion that `ObjectReference` should not be used?

It's pretty strong.  We should not use ObjectReference.  

```
//  1. Ignored fields.  It includes many fields which are not generally honored.  For instance, ResourceVersion and FieldPath are both very rarely valid in actual usage.
//  2. Invalid usage help.  It is impossible to add specific help for individual usage.  In most embedded usages, there are particular
//     restrictions like, "must refer only to types A and B" or "UID not honored" or "name must be restricted".
//     Those cannot be well described when embedded.
//  3. Inconsistent validation.  Because the usages are different, the validation rules are different by usage, which makes it hard for users to predict what will happen.
//  4. The fields are both imprecise and overly precise.  Kind is not a precise mapping to a URL. This can produce ambiguity
//     during interpretation and require a REST mapping.  In most cases, the dependency is on the group,resource tuple
//     and the version of the actual struct is irrelevant.
//  5. We cannot easily change it.  Because this type is embedded in many locations, updates to this type
//     will affect numerous schemas.  Don't make new APIs embed an underspecified API type they do not control.
```